### PR TITLE
campaigns: Add WebhookURL to ExternalService

### PR DESCRIPTION
**Pull request by ryanslade**
The CampaignWebhookURL optional string is now included on ExternalService in the API. This can be used from the frontend to show a URL that the admin should configure their webook to point at.

Additionally, we now check for this parameter when parsing both GitHub and BitbucketServer webhooks. 

Part of: https://github.com/sourcegraph/sourcegraph/issues/9008

### CODE REVIEW

File: cmd/frontend/graphqlbackend/external_service.go [hunk 1]
```diff
@@ -74,6 +82,34 @@ func (r *externalServiceResolver) UpdatedAt() DateTime {
 	return DateTime{Time: r.externalService.UpdatedAt}
 }
 
+func (r *externalServiceResolver) CampaignWebhookURL() *string {
+	r.webhookURLOnce.Do(func() {
+		parsed, err := extsvc.ParseConfig(r.externalService.Kind, r.externalService.Config)
+		if err != nil {
```
Comment (mrnugget) [inline cmd/frontend/graphqlbackend/external_service.go:88]: Why not return this error? We use the same pattern (`sync.Once`, value field, error field) in other resolvers. Example:

https://github.com/sourcegraph/sourcegraph/blob/91a98b8aad34aa5642222088efaa45756b237865/cmd/frontend/graphqlbackend/repository_comparison.go#L225-L226

In this case the conversion would not be too hard and it would lead to one less error being swallowed somewhere :)

**Comment by ryanslade** [comment]
Well, the TS code isn't compiling so will have to make the change in this PR :)

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 2]
```diff
@@ -1560,6 +1560,8 @@ type ExternalService implements Node {
     createdAt: DateTime!
     # When the external service was last updated.
     updatedAt: DateTime!
+    # An optional url that will be populated when a campaign webhook has been configured
+    campaignWebhookURL: String
```
Comment (mrnugget) [inline cmd/frontend/graphqlbackend/schema.graphql:1564]: Does it need to be campaign specific? The rest of the documentation related to webhooks doesn't mention campaigns, except to say "this is currently only used by campaigns and we recommend setting it up", but the idea is that maybe we can use it for more things in the future (permissions, repositories, etc.)

```suggestion
    # An optional url that will be populated when a webhook has been configured for this external service.
    webhookURL: String
```

File: enterprise/internal/campaigns/webhooks.go [hunk 3]
```diff
@@ -253,8 +253,22 @@ func (h *GitHubWebhook) parseEvent(r *http.Request) (interface{}, *repos.Externa
 
 	sig := r.Header.Get("X-Hub-Signature")
 
+	rawID := r.FormValue(campaigns.ExternalServiceIDParam)
+	var externalServiceID int64
+	// Older hook URL's may not contain the external service
```
Comment (mrnugget) [inline enterprise/internal/campaigns/webhooks.go:258]:
```suggestion
	// If a webhook was setup before we introduced the externalServiceID as part of the URL,
	// the webhook requests may not contain the external service ID, so we need to fall back.
```

File: internal/campaigns/types.go [hunk 4]
```diff
@@ -1270,3 +1271,12 @@ type ChangesetSyncData struct {
...
 	return time.Unix(0, ms*int64(time.Millisecond))
 }
+
+const ExternalServiceIDParam = "externalServiceID"
+
+func WebhookURL(externalServiceID int64) string {
+	host := conf.Cached(func() interface{} {
+		return conf.Get().ExternalURL
+	})().(string)
+	return fmt.Sprintf("https://%s/?%s=%d", host, ExternalServiceIDParam, externalServiceID)
```
Comment (mrnugget) [inline internal/campaigns/types.go:1281]: 1. Doesn't this need to include the path? The [GitHub docs](https://docs.sourcegraph.com/admin/external_service/github#webhooks) mention `/.api/github-webhooks` as the path. And the [bitbucket-server-plugin README](https://github.com/sourcegraph/bitbucket-server-plugin/blob/master/src/main/java/com/sourcegraph/webhook/README.md) mentions `/.api/bitbucket-server-webhooks`.
2. Can we _assume_ `https` here? Or should we use the protocol from `ExternalURL`?

File: internal/extsvc/types.go [hunk 5]
```diff
@@ -70,3 +74,27 @@ type AccountID string
 // code host. It can be the string representation of an integer (e.g. GitLab and Bitbucket
 // Server) or a GraphQL ID (e.g. GitHub) depends on the code host type.
 type RepoID string
+
+// ParseConfig attempts to parse external service config into a strong typed
```
Comment (mrnugget) [inline internal/extsvc/types.go:78]:
```suggestion
// ParseConfig attempts to unmarshal the given JSON config into a configuration struct defined in the schema package.
```

I don't think we should say "strongly typed" when we return an `interface{}` 😄

File: internal/campaigns/types.go [hunk 6]
```diff
@@ -1270,3 +1271,12 @@ type ChangesetSyncData struct {
 func unixMilliToTime(ms int64) time.Time {
 	return time.Unix(0, ms*int64(time.Millisecond))
 }
+
+const ExternalServiceIDParam = "externalServiceID"
+
+func WebhookURL(externalServiceID int64) string {
```
Comment (mrnugget) [inline internal/campaigns/types.go:1277]: Any reason for why is this here in the `campaigns` package, but the `ParseConfig` is in the `extsvc` package? I'd expect this to be in `extsvc` too, since it's pretty generic and doesn't contain any information that the external service schema/config doesn't contain.

Comment (ryanslade) [reply internal/campaigns/types.go:1277]: That was there because I thought it was campaign specific, I'll move it

**Comment by ryanslade** [comment]
@mrnugget PTAL

File: internal/extsvc/types.go [hunk 7]
```diff
@@ -70,3 +75,44 @@ type AccountID string
...
+		return conf.Get().ExternalURL
+	})().(string)
+	var path string
+	switch strings.ToLower(kind) {
+	case "github":
+		path = "github-webhooks"
+	case "bitbucketserver":
+		path = "bitbucket-server-webhooks"
+	default:
+		return "", fmt.Errorf("uknown external service kind: %q", kind)
```
Comment (eseliger) [inline internal/extsvc/types.go:115]:
```suggestion
		return "", fmt.Errorf("unknown external service kind: %q", kind)
```

Comment (mrnugget) [reply internal/extsvc/types.go:115]: I feel like @unknwon snuck this in here 😄

File: cmd/frontend/graphqlbackend/external_service.go (hunk 1 above)
Comment (mrnugget) [reply cmd/frontend/graphqlbackend/external_service.go:88]: Ping on this @ryanslade :)

File: cmd/frontend/graphqlbackend/schema.graphql (hunk 2 above)
Comment (mrnugget) [reply cmd/frontend/graphqlbackend/schema.graphql]: You marked the conversation as resolved, but I still see `campaignWebhookURL`. Does that mean you don't want to rename it? Little note appreciated for the next round of reviews.

File: cmd/frontend/graphqlbackend/external_service.go (hunk 1 above)
Comment (mrnugget) [reply cmd/frontend/graphqlbackend/external_service.go:88]: Ignore the ping, sorry. I had the wrong view open in GitHub.

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 8]
```diff
@@ -1560,6 +1560,8 @@ type ExternalService implements Node {
     createdAt: DateTime!
     # When the external service was last updated.
     updatedAt: DateTime!
+    # An optional url that will be populated when webhooks have been configured
```
Comment (mrnugget) [inline cmd/frontend/graphqlbackend/schema.graphql:1563]:
```suggestion
    # An optional url that will be populated when webhooks have been configured for the external service.
```

File: enterprise/internal/campaigns/webhooks.go [hunk 9]
```diff
@@ -824,7 +836,10 @@ func (h *BitbucketServerWebhook) syncWebhook(externalServiceID int64, con *schem
 	}
 
 	// Secret has changed to a non blank value, upsert
-	endpoint := fmt.Sprintf("%s/.api/bitbucket-server-webhooks?%s=%d", externalURL, externalServiceIDParam, externalServiceID)
+	endpoint, err := extsvc.WebhookURL("bitbucketserver", externalServiceID)
```
Comment (mrnugget) [inline enterprise/internal/campaigns/webhooks.go:839]: How about using the `bitbucketserver.ServiceType` and `github.ServiceType` constants?

File: enterprise/internal/campaigns/webhooks.go [hunk 10]
```diff
@@ -881,7 +896,7 @@ func (h *BitbucketServerWebhook) parseEvent(r *http.Request) (interface{}, *repo
 
 	sig := r.Header.Get("X-Hub-Signature")
 
-	rawID := r.FormValue(externalServiceIDParam)
+	rawID := r.FormValue(extsvc.ExternalServiceIDParam)
```
Comment (mrnugget) [inline enterprise/internal/campaigns/webhooks.go:899]: Reading it in this context, I think adding "Webhook" to the name might make things clearer:

```suggestion
	rawID := r.FormValue(extsvc.WebhookParam)
```

**Comment by eseliger** [comment]
I will add the UI in a follow-up PR as I don't know if I find the time today or so, to unblock this I opened: https://github.com/sourcegraph/sourcegraph/issues/10300

File: internal/extsvc/types.go [hunk 11]
```diff
@@ -70,3 +75,44 @@ type AccountID string
...
+	}
+	return fmt.Sprintf("https://%s/%s?%s=%d", host, path, ExternalServiceIDParam, externalServiceID), nil
```
Comment (mrnugget) [inline internal/extsvc/types.go:117]: This still needs the `/.api` part, right? And it also doesn't address part 2 of my comment here: https://github.com/sourcegraph/sourcegraph/pull/10292#discussion_r418007448

File: enterprise/internal/campaigns/webhooks.go (hunk 10 above)
Comment (ryanslade) [reply enterprise/internal/campaigns/webhooks.go:899]: It's specific to the ID, so I think as it's in the extsvc package already I'll call it `extsvc.IDParam`

File: internal/extsvc/types.go (hunk 11 above)
Comment (ryanslade) [reply internal/extsvc/types.go:117]: Sorry about that. Fixed now and confirmed by querying the GraphQL api

File: cmd/frontend/graphqlbackend/external_service.go [hunk 12]
```diff
@@ -74,6 +82,38 @@ func (r *externalServiceResolver) UpdatedAt() DateTime {
...
+				r.webhookURL = u
+			}
+		case *schema.GitHubConnection:
+			if len(c.Webhooks) > 0 {
+				r.webhookURL = u
+			}
+		}
+	})
+	if r.webhookURL == "" {
+		return nil, r.webhookErr
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/external_service.go:112]: This means the caller would get a `nil` error and a `nil` pointer to the webhook URL. So the question is: is it guaranteed the caller would only call this method for GitHub and Bitbucket Server? (despite we only support these two code hosts currently IIRC)

File: cmd/frontend/graphqlbackend/schema.graphql [hunk 13]
```diff
@@ -1560,6 +1560,8 @@ type ExternalService implements Node {
     createdAt: DateTime!
     # When the external service was last updated.
     updatedAt: DateTime!
+    # An optional url that will be populated when webhooks have been configured for the external service.
```
Comment (unknwon) [inline cmd/frontend/graphqlbackend/schema.graphql:1563]: nit:
```suggestion
    # An optional URL that will be populated when webhooks have been configured for the external service.
```

File: enterprise/internal/campaigns/webhooks.go [hunk 14]
```diff
@@ -59,11 +59,11 @@ func (h Webhook) getRepoForPR(
 		},
 	})
 	if err != nil {
-		return nil, errors.Wrap(err, "Failed to load repository")
+		return nil, errors.Wrap(err, "failed to load repository")
```
Comment (unknwon) [inline enterprise/internal/campaigns/webhooks.go:62]: 👍

File: internal/extsvc/types.go [hunk 15]
```diff
@@ -70,3 +75,45 @@ type AccountID string
...
+	}
+	return cfg, jsonc.Unmarshal(config, cfg)
+}
+
+const IDParam = "externalServiceID"
+
+func WebhookURL(kind string, externalServiceID int64) (string, error) {
+	host := conf.Cached(func() interface{} {
+		return conf.Get().ExternalURL
+	})().(string)
```
Comment (unknwon) [inline internal/extsvc/types.go:107]: I'm not 100% sure we should put this blocking logic in `extsvc` package, the `internal/conf/computed.go` or `cmd/frontend/globals` is probably a more appropriate package you're looking for.

Comment (mrnugget) [reply internal/extsvc/types.go:107]: Agree! Those are good suggestions for this piece of code.

File: cmd/frontend/graphqlbackend/external_service.go (hunk 12 above)
Comment (ryanslade) [reply cmd/frontend/graphqlbackend/external_service.go:112]: This is in a graphql resolver as an optional string so I think it's fine to return nil here.

File: internal/extsvc/types.go (hunk 15 above)
Comment (ryanslade) [reply internal/extsvc/types.go]: Is this what you mean? https://github.com/sourcegraph/sourcegraph/pull/10292/commits/5c0408767377104aaf9e2c56021eafd408b538c3

Comment (unknwon) [reply internal/extsvc/types.go]: @ryanslade LGTM 👍 One last nitty-picky: could we pass in the external URL to `WebhookURL` function? That would make our lives easier to write tests for `WebhookURL` function (no need to mock conf).

Comment (ryanslade) [reply internal/extsvc/types.go]: Ok, will do
